crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -189,8 +189,17 @@ impl ClassInfoConstraintFunction {
                     Some(constraint_fn(class_literal.default_specialization(db)))
                 }
             }
-            Type::SubclassOf(subclass_of_ty) => {
-                subclass_of_ty.subclass_of().into_class().map(constraint_fn)
+            Type::SubclassOf(subclass_of_ty) => match subclass_of_ty.subclass_of() {
+                SubclassOfInner::Class(class) => Some(constraint_fn(class)),
+                SubclassOfInner::Dynamic(dynamic) => Some(Type::Dynamic(dynamic)),
+            },
+            Type::Dynamic(_) => Some(classinfo),
+            Type::Intersection(intersection) if intersection.negative(db).is_empty() => {
+                let mut builder = IntersectionBuilder::new(db);
+                for element in intersection.positive(db) {
+                    builder = builder.add_positive(self.generate_constraint(db, *element)?);
+                }
+                Some(builder.build())
```

Comment:
Can you explain the reasoning behind this branch? Why or why not would something similar apply to `Type::Union`?

Reply:
Yeah, I think you're both right!

I was looking at it pretty narrowly through the lens of "what would be required to respect the gradual guarantee here" -- and it seemed obvious that, to respect the gradual guarantee, if we intersected with `Any` when the second argument to `isinstance()` was of type `Any` and we intersected with `int` when the second argument was of type `type[int]`, we should therefore also intersect with `int & Any` when the second argument was of `type[int] & Any`.

But if we're making these changes to intersections, then it does definitely make sense to make equivalent changes to unions -- it's sound behaviour and consistent behaviour, even if it's not strictly "necessary" to support the gradual guarantee.